### DISCUSSION TITLE
Optionally suppress webdriver for API testing

### DISCUSSION THREAD

**Discussion by tzurp** [webdriverio/webdriverio#8016] [Community]
Most of our WebderiveIO UI tests are deeply coupled to REST API calls. We use REST API for test setup and for data verification.
Since we already have the REST API infrastructure it is just natural to use our framework for running pure REST API tests. The problem is that we can't find a way do disable the browser (and the webdriver) initialization. Is there an option to suppress the webdriver? If not it seems to be a much needed feature because it will make wdio a much flexible and comprehensive testing framework.

**Comment by christian-bromann** [comment]
> Is there an option to suppress the webdriver?

There isn't. WebdriverIO is designed for pure e2e testing. If you want to run API tests use Mocha, Jasmine and Cucumber directly depending on what you use in combination with WebdriverIO.

> If not it seems to be a much needed feature because it will make wdio a much flexible and comprehensive testing framework.

It would also broaden the scope of the framework. A lot of things WebdriverIO does to simplify your UI tests is not needed for API testing. I understand the desire to use one framework for everything but that is not realistic in reality. You also don't use a hammer as single tool to build a whole house.

That said, I was open to the idea of introducing API testing to WebdriverIO in the past and continue stay open to this idea. However it requires a design proposal that goes a bit beyond the "not starting a browser" tweak. If you are interested working on such a proposal and committed to contribute the work, I am happy to chat about it.

**Reply by tzurp** [reply]
@christian-bromann before we enter to design proposals we must understand the organizational problems that arises when using wdio. First we need to agree that to write proper e2e UI tests, in many cases there is a need to combine API calls in the test scenario for test setup and verifications. From **organizational** point of view it means that the infrastructure for the API testing is immersed in the UI testing framework. Now, if we need to run pure API tests for the same products, and we need to use for that a separate framework, it would result with a huge waste of resources due to duplication of the API code. I think that for some it might be a deal breaker. Once we agree about the organizational needs, we can step forward for the design proposals.

**Reply by christian-bromann** [reply]
> Now, if we need to run pure API tests for the same products, and we need to use for that a separate framework, it would result with a huge waste of resources due to duplication of the API code.

I am not sure why API code has to get duplicated for this. Instead of running API tests with WebdriverIO you can run them via plain Mocha, Jasmine or Cucumber. Given that API tests are faster than UI tests (generally, not always) I would run them before kicking of the UI tests. At the end of the day your pipeline could look as follows:

```
- name: API tests
   run: npx mocha ./tests/api
- name: UI tests
   run: npx wdio wdio.conf.js
```
